src/Compilers/Core/Portable/Diagnostic/CommonMessageProvider.cs

```diff
@@ -99,9 +104,13 @@ public Diagnostic CreateDiagnostic(int code, Location location)
         /// <summary>
         /// Given an error code (like 1234) return the identifier (CS1234 or BC1234).
         /// </summary>
+        [PerformanceSensitive(
+            "https://github.com/dotnet/roslyn/issues/31964",
+            AllowCaptures = false,
+            Constraint = "Frequently called by error list filtering; avoid allocations")]
         public string GetIdForErrorCode(int errorCode)
         {
-            return CodePrefix + errorCode.ToString("0000");
+            return _errorIdCache.GetOrAdd((CodePrefix, errorCode), key => key.prefix + key.code.ToString("0000"));
```

Comment:
Consider caching this lambda. If we're doing this because perf, we should make sure we're not replacing lots of `string` allocations with lots of delegate allocations.

Reply:
> Consider caching this lambda. 

The lambda has no captures. It is cached by the compiler.

Reply:
Ah, so it is.